File: Data Structures & Algorithms/minimum-stack/submission-1.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.minimum = math.inf
        self.min_count = 0

    def push(self, val: int) -> None:
        self.stack.append(val)
        if val<self.minimum:
            self.minimum = val
            self.min_count = 1
        elif val==self.minimum:
            self.min_count+=1

    def pop(self) -> None:
        if len(self.stack)==0: return
        if self.stack[-1] == self.minimum:
            self.min_count-=1
        self.stack = self.stack[:-1]

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        if self.min_count<=0:
            self.minimum = self.stack[0]
            for i in range(len(self.stack)):
                if self.stack[i]<self.minimum:
                    self.minimum = self.stack[i]
        return self.minimum

File: Data Structures & Algorithms/validate-parentheses/submission-1.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        map<char,char> corr = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };

        for(char c : s) {
            if(c=='(' || c=='[' || c=='{') st.push(c);
            else if(st.empty() || st.top() != corr[c]) return false;
            else st.pop();
        }

        return st.empty();
    }
};
